gen_program_tests getting killed almost consistently

Something is broken with one of the integration tests on master, but only in CI.
Several otherwise green PRs are failing often on one of the pkg test shards.
It appears that the test gets killed after some time of posting heartbeats.

```
💓
💓
make: *** [Makefile:139: gotestsum/pkg] Terminated
Error: Process completed with exit code 143.
```

The failure is flaky but it's been leaning more and more red recently.
The cause is unclear because nothing is posted about it,
but one of the ongoing guesses is that
we're hitting a memory usage limit in one of those tests,
which causes a termination.

Examples:
- https://github.com/pulumi/pulumi/actions/runs/3962273901/jobs/6789198464
- https://github.com/pulumi/pulumi/actions/runs/3962096772/jobs/6788386893

The relevant command (with the same PKGs and tags) succeeds on local executions.

This is not just a problem in PRs. bors also fails with these occasionally:

- https://app.bors.tech/batches/286903
- https://app.bors.tech/batches/286697

--- a/pkg/codegen/testing/test/program_driver.go
+++ b/pkg/codegen/testing/test/program_driver.go
@@ -381,7 +381,10 @@ func TestProgramCodegen(
 	for _, tt := range testcase.TestCases {
 		tt := tt // avoid capturing loop variable
 		t.Run(tt.Directory, func(t *testing.T) {
-			t.Parallel()
+			// These tests should not run in parallel.
+			// They take up a fair bit of memory
+			// and can OOM in CI with too many running.
+
 			var err error
 			if tt.Skip.Has(testcase.Language) {
 				t.Skip()

